### PR TITLE
types: Fix "Invalid time format" caused by daily saving time

### LINKED ISSUE #13605: "Invalid time format" caused by time zone issue
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
I am in the `-8:00` time zone.
```
➜  work date -R
Tue, 19 Nov 2019 15:51:45 -0800
```
With the following operation, I meet the invalid time format issue.
```
use test;
set @@time_zone = '+0:00';
SELECT @@global.time_zone, @@session.time_zone;
create table t (ts timestamp);
insert into t values ("2018-03-11 02:00:16");
```

2. What did you expect to see?
```
Query OK, 1 row affected (0.01 sec)
```

3. What did you see instead?
```
ERROR 1292 (22007): invalid time format: '{2018 3 11 2 0 16 0}'
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
| Release Version: v4.0.0-alpha-921-g77faf6466-dirty
Git Commit Hash: 77faf64668213114ea33f565c84dde52a5359667
Git Branch: master
UTC Build Time: 2019-11-19 11:23:59
GoVersion: go version go1.13.4 darwin/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Deardrops**
### What problem does this PR solve? 

Fix #13605 

### What is changed and how it works?

Some regions use [daylight saving time](https://en.wikipedia.org/wiki/Daylight_saving_time).

In MySQL, it will throw the error "Invalid time format" when use the `datetime` between the advanced hour of daylight saving time (e.g. 2018/3/11 2:00~3:00 in Los Angeles). Actually, if the `@@time_zone` sql variable is set to region where uses daylight saving time, the error occurred. But if the `@@time_zone` is other region, no error occurred, even if the system time zone is the region where uses daylight saving time.

But TiDB will throw an unexpected error "Invalid time format" when the timezone of system is in region where use daylight saving time.

**Fix method:**

When checking the time, we use the SQL variable `@@time_zone` instead of directly use the timezone of system.

In details, In `checkTimestampType` function, we use the `TimeZone` of `statementContext`, instead of local timezone provided by Golang.

### Check List 

Tests 

 - Unit test

Code changes

 - None

Side effects

 - None

Related changes

 - None

Release note

 - None

### CODE REVIEW

File: types/time_test.go [hunk 1]
```diff
@@ -1101,11 +1101,59 @@ func (s *testTimeSuite) TestCheckTimestamp(c *C) {
...
 		}
 	}
+
+	// Issue #13605: "Invalid time format" caused by time zone issue
+	// Some regions like Los Angeles use daylight saving time, see https://en.wikipedia.org/wiki/Daylight_saving_time
+	losAngelesTz, _ := time.LoadLocation("America/Los_Angeles")
+	rawLocal := time.Local
+	time.Local = losAngelesTz // Hack Local timezone
+
+	tests = []struct {
```
Comment (bb7133) [inline types/time_test.go]: How about adding a new test for London DST?

> March 31, 2019, 1:00:00 am clocks were turned forward 1 hour to March 31, 2019, 2:00:00 am local daylight time instead.

Comment (Deardrops) [reply types/time_test.go]: @bb7133 comment addressed, PTAL

**Comment by bb7133** [comment]
LGTM

**Pull request merged**